src/nomad_simulations/properties/band_structure.py

```diff
@@ -0,0 +1,370 @@
+        """
+        if self.occupation is None or len(self.occupation) == 0:
+            logger.warning('Cannot find `occupation` defined.')
+            return False
+        if self.value is not None and self.value.shape != self.occupation.shape:
+            logger.warning(
+                'The shape of `value` and `occupation` are different. They should have the same shape.'
+            )
+            return False
+        return True
+
+    def order_eigenvalues(
+        self, logger: BoundLogger
+    ) -> Tuple[Optional[pint.Quantity], Optional[np.ndarray]]:
+        """
+        Order the eigenvalues based on the `value` and `occupation`. The return `value` and
+        `occupation` are flattened.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (Tuple[Optional[pint.Quantity], Optional[list]]): The flattened and sorted `value` and `occupation`.
+        """
+
+        # Check if `value` exists
+        if self.value is None or len(self.value) == 0:
+            logger.error('Could not find `value` defined.')
+            return None, None
+
+        # Check if `occupation` exists and have the same shape as `value`
+        if not self._check_occupation_shape(logger):
+            return None, None
+        total_shape = np.prod(self.value.shape)
+
+        # Order the indices in the flattened list of `value`
+        flattened_value = self.value.reshape(total_shape)
+        flattened_occupation = self.occupation.reshape(total_shape)
+        sorted_indices = np.argsort(flattened_value, axis=0)
+
+        sorted_value = (
+            np.take_along_axis(flattened_value.magnitude, sorted_indices, axis=0)
+            * flattened_value.u
+        )
+        sorted_occupation = np.take_along_axis(
+            flattened_occupation, sorted_indices, axis=0
+        )
+        self.m_cache['sorted_eigenvalues'] = True
+        return sorted_value, sorted_occupation
+
+    def resolve_homo_lumo_eigenvalues(
+        self, logger: BoundLogger
+    ) -> Tuple[Optional[pint.Quantity], Optional[pint.Quantity]]:
+        """
+        Resolve the `highest_occupied` and `lowest_unoccupied` eigenvalues by performing a binary search on the
+        flattened and sorted `value` and `occupation`. If these quantities already exist, overwrite them or return
+        them if it is not possible to resolve from `value` and `occupation`.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (Tuple[Optional[pint.Quantity], Optional[pint.Quantity]]): The `highest_occupied` and
+            `lowest_unoccupied` eigenvalues.
+        """
+        # Sorting `value` and `occupation`
+        sorted_value, sorted_occupation = self.order_eigenvalues(logger)
+        if sorted_value is None or sorted_occupation is None:
+            if self.highest_occupied is not None and self.lowest_unoccupied is not None:
```

Comment:
are you not supposed to return right away self.highest(lowest)(un)occupied if they are assigned?